pkg/devfile/resources.go

```diff
@@ -0,0 +1,258 @@
+
+				dockerImagePort := devfileObj.Data.GetMetadata().Attributes.GetString("alpha.dockerimage-port", &err)
+				if err != nil {
+					errMsg := fmt.Sprintf("Failed to get the Dockerfile location from devfile metadata attribute 'alpha.build-dockerfile': %v", err)
+					klog.Error(errMsg)
+
+					return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+				}
+
+				imageComponentFilter := common.DevfileOptions{
+					ComponentOptions: common.ComponentOptions{
+						ComponentType: devfilev1.ImageComponentType,
+					},
+				}
+
+				// Service
+				serviceResource, err := GetService(devfileObj, imageComponentFilter, dockerImagePort)
+				if err != nil {
+					errMsg := fmt.Sprintf("Failed to get service for the devfile: %v", err)
+					klog.Error(errMsg)
+					return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+				}
+				appendedResources.Services = append(appendedResources.Services, serviceResource)
+
+				// Route
+				routeResource := GetRouteForDockerImage(name, dockerImagePort, "/", false, nil)
+				appendedResources.Routes = append(appendedResources.Routes, routeResource)
+
+				returnDeploymentOnly = false
+
+				// break here as we are no longer interested in parsing a YAML file
+				break
+			}
+
+			fs := afero.Afero{Fs: afero.NewOsFs()}
+			values, err := parser.ReadKubernetesYaml(src, fs)
+			if err != nil {
+				errMsg := fmt.Sprintf("Failed to read the Kubernetes yaml from devfile: %v", err)
+				klog.Error(errMsg)
+				return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+			}
+
+			resources, err := parser.ParseKubernetesYaml(values)
+			if err != nil {
+				errMsg := fmt.Sprintf("Failed to parse the Kubernetes yaml data from devfile: %v", err)
+				klog.Error(errMsg)
+				return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+			}
+
+			for _, endpoint := range component.Kubernetes.Endpoints {
+				if endpoint.Exposure != devfilev1.NoneEndpointExposure {
+					var isSecure bool
+					if endpoint.Secure != nil {
+						isSecure = *endpoint.Secure
+					}
+					resources.Routes = append(resources.Routes, GetRouteForDockerImage(name, fmt.Sprintf("%d", endpoint.TargetPort), endpoint.Path, isSecure, endpoint.Annotations))
+				}
+			}
+
+			appendedResources.Deployments = append(appendedResources.Deployments, resources.Deployments...)
+			appendedResources.Services = append(appendedResources.Services, resources.Services...)
+			appendedResources.Routes = append(appendedResources.Routes, resources.Routes...)
+		}
+	}
+
+	var errMsg string
+
+	if len(appendedResources.Deployments) > 0 {
+		deployment = appendedResources.Deployments[0]
+	} else {
+		errMsg = "no deployment definition was found in the devfile sample"
+	}
+
+	if len(appendedResources.Services) > 0 {
+		service = appendedResources.Services[0]
+	} else if !returnDeploymentOnly {
+		if len(errMsg) != 0 {
+			errMsg = errMsg + ", "
+		}
+		errMsg = errMsg + "no service definition was found in the devfile sample"
```

Comment:
instead of doing the concatenating. can use `github.com/hashicorp/go-multierror`?
and 
```
errMsg = multierror.Append(errMsg, fmt.Errorf("no service definition was found in the devfile sample"))
```

Reply:
oh should have used this